### PR TITLE
Add common code that is shared between SEV keep and shim

### PR DESCRIPTION
**Pull request by connorkuehl**
This will allow us to more easily parallelize our efforts to porting in Harald's work. There will be two projects consuming this common core:

- 'vmrun' AKA enarx-keep-sev
- 'kernel' AKA enarx-keep-sev-shim

I'd prefer we err on opening debt issues than making any large or invasive changes. This is the beginning of a rather delicate porting of a lot of SEV code.

### CODE REVIEW

File: vmsyscall/src/memory_map.rs [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    Bootloader,
+    /// Frame at address zero.
+    ///
+    /// (shouldn't be used because it's easy to make mistakes related to null pointers)
+    FrameZero,
+    /// An empty region with size 0
+    Empty,
+    /// Additional variant to ensure that we can add more variants in the future without
+    /// breaking backwards compatibility.
+    NonExhaustive,
```
Comment (haraldh) [inline vmsyscall/src/memory_map.rs]: why did you remove `#[doc(hidden)]` ??

Comment (connorkuehl) [reply vmsyscall/src/memory_map.rs]: No good reason. Just saw the comment above it and thought it'd be useful information to appear in the generated documentation.

File: vmsyscall/src/lib.rs [hunk 2]
```diff
@@ -1,94 +1,98 @@
...
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq)]
 pub enum Error {
-    /// standard error
     Errno(i64),
-    /// serialize error
     SerializeError,
-    /// deserialize error
     DeSerializeError,
 }
```
Comment (npmccallum) [inline vmsyscall/src/lib.rs]: I don't think Ser/DeSer errors are used anymore. Which also raises the question about whether this enum is needed.

File: vmsyscall/src/memory_map.rs (hunk 1 above)
Comment (npmccallum) [reply vmsyscall/src/memory_map.rs]: Is there a reason we can't use `#![non_exhaustive]` here?

File: vmsyscall/src/memory_map.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    /// Returns the physical start address of the memory region.
+    pub fn start_addr(&self) -> u64 {
+        self.start_frame_number * PAGE_SIZE
+    }
+
+    /// Returns the physical end address of the memory region.
+    pub fn end_addr(&self) -> u64 {
+        self.end_frame_number * PAGE_SIZE
+    }
+}
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: I think we should use `core::ops::Range` here instead. This is also isomorphic with `span::Span`. Make a newtype `Frame(u64)` with conversions to/from addresses. Then just do `Range<Frame>` or `Span<Frame>`.

File: vmsyscall/src/memory_map.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    /// Mark a region as usable.
+    pub fn set_region_type_usable(&mut self, region_type: MemoryRegionType) {
+        self.iter_mut().for_each(|r| {
+            if r.region_type == region_type {
+                r.region_type = MemoryRegionType::Usable
+            }
+        });
+    }
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: This seems like this could be made more generic as `fn set_region_type(rtype: MemoryRegionType, whentype: Option<MemoryRegionType>)`.

File: vmsyscall/src/memory_map.rs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                && last_region.range.end_frame_number >= region.range.start_frame_number
+                && last_region.range.end_frame_number <= region.range.end_frame_number
+            {
+                last_region.range.end_frame_number = region.range.end_frame_number;
+                return Err(());
+            }
+            Ok(())
+        }) {
+            return;
+        };
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: I think this is more clearly expressed as a `for` loop where you test for exclusions and then `continue`. If you bypass all the exclusion tests then you modify and return.

File: vmsyscall/src/memory_map.rs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    /// Mark a region as allocated.
+    pub fn mark_allocated_region(&mut self, region: MemoryRegion) {
+        let mut region = region;
+        for r in self.iter_mut() {
+            // New region inside region of same type
+            if r.region_type == region.region_type
+                && r.range.start_frame_number <= region.range.start_frame_number
+                && r.range.end_frame_number >= region.range.end_frame_number
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: This pattern or similar occurs frequently. Could we make this an operation on the `MemoryRegion` struct?

File: vmsyscall/src/memory_map.rs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                panic!(
+                    "region {:x?} overlaps with non-usable region {:x?}",
+                    region, r
+                );
+            }
+
+            match region
+                .range
+                .start_frame_number
+                .cmp(&r.range.start_frame_number)
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: I strongly dislike wraps in a match statement. Please use a temporary variable.

Comment (npmccallum) [reply vmsyscall/src/memory_map.rs]: Also, this is screaming for an operation on the `Frame` newtype (see below).

File: vmsyscall/src/memory_map.rs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            {
+                Ordering::Equal => {
+                    if region.range.end_frame_number < r.range.end_frame_number {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // ----RRRR-----------
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: These comments are as good as ciphertext to me. I don't understand them.

File: vmsyscall/src/memory_map.rs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    pub fn set_region_type_usable(&mut self, region_type: MemoryRegionType) {
+        self.iter_mut().for_each(|r| {
+            if r.region_type == region_type {
+                r.region_type = MemoryRegionType::Usable
+            }
+        });
+    }
+
+    /// Add a region to the `MemoryMap`.
+    pub fn add_region(&mut self, region: MemoryRegion) {
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: I think `add_region()` can be private.

File: vmsyscall/src/memory_map.rs [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+                    .start_frame_number
+                    .cmp(&r2.range.start_frame_number);
+
+                if ordering == Ordering::Equal {
+                    r1.range.end_frame_number.cmp(&r2.range.end_frame_number)
+                } else {
+                    ordering
+                }
+            }
+        });
```
Comment (npmccallum) [inline vmsyscall/src/memory_map.rs]: This would be far more readable if you had a `Frame` newtype that implemented `PartialOrd` and then implemented `PartialOrd` on `Region`.

File: vmsyscall/src/memory_map.rs (hunk 1 above)
Comment (haraldh) [reply vmsyscall/src/memory_map.rs]: > Is there a reason we can't use `#![non_exhaustive]` here?

probably not

File: vmsyscall/src/lib.rs (hunk 2 above)
Comment (haraldh) [reply vmsyscall/src/lib.rs]: correct

File: vmsyscall/src/bootinfo.rs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+//! the `BootInfo` structure which is handed to the kernel.
+//!
+//! Initially copied from:
+//! https://github.com/rust-osdev/bootloader/blob/90f5b8910d146d6d489b70a6341d778253663cfa/src/bootinfo/mod.rs
+
+use crate::memory_map::MemoryMap;
+use core::fmt;
+
+/// Hard-coded trigger port
+pub const SYSCALL_TRIGGER_PORT: u16 = 0xFF;
```
Comment (haraldh) [inline vmsyscall/src/bootinfo.rs]: oh, we can move that into `vmrun` ... doesn't have to be here anymore, because it's a BootInfo argument

File: vmsyscall/src/bootinfo.rs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// Elf program headers of the ring3 executable
+    pub load_addr: *const u8,
+
+    /// Elf number of program headers of the ring3 executable
+    pub elf_phnum: usize,
+
+    /// Syscall trigger port
+    pub syscall_trigger_port: u16,
+}
```
[the remark this answers is not in the document]
Comment (haraldh) [reply vmsyscall/src/bootinfo.rs]: Since it's part of the memory now (vmsyscall page) , wouldn't it be measured like the other parts?

File: vmsyscall/src/memory_map.rs (hunk 9 above)
Comment (haraldh) [reply vmsyscall/src/memory_map.rs]: Why? I would like to use it.

File: vmsyscall/src/memory_map.rs (hunk 8 above)
Comment (haraldh) [reply vmsyscall/src/memory_map.rs]: These are both regions, which start at the same point.. one is bigger than the other.

File: vmsyscall/src/memory_map.rs (hunk 9 above)
Comment (npmccallum) [reply vmsyscall/src/memory_map.rs]: Sure, no problem. I just didn't see any immediate reference to it outside the context.

File: vmsyscall/src/memory_map.rs (hunk 4 above)
Comment (npmccallum) [reply vmsyscall/src/memory_map.rs]: The `Option` might not make sense depending on how this is used.

File: vmsyscall/src/bootinfo.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply vmsyscall/src/bootinfo.rs]: Probably. I'm just noting that it is critical that the host cannot choose the starting point for the guest without it being measured. A comment here could be a helpful warning to future coders.

File: vmsyscall/src/memory_map.rs (hunk 4 above)
Comment (haraldh) [reply vmsyscall/src/memory_map.rs]: Right.. but the generic use might not make sense. At least I can't think of any currently.

**Comment by connorkuehl** [comment]
Most recent push:

- [x] Remove `SYSCALL_TRIGGER_PORT` const as it will be in `enarx-keep-sev` instead.
- [x] Remove wrapping match statement
- [x] Simplify seek operation in `add_region`
- [x] Use `#[non_exhaustive]` on enum
- [x] Remove now-unused Error enum
- [x] Remove Memory prefix from types in `memory_map`
- [x] Implement Ord and PartialOrd on the memory `Region` type to simplify sorting... stepping stone toward Nathaniel's feedback based on the Frame newtype.
- [ ] Use Span/Range instead of FrameRange (started to spiral a bit in complexity)
- [ ] Frame newtype
- [ ] Make `set_region_type` more generic

I'm still working on the changes that aren't checked off above.

**Comment by connorkuehl** [comment]
Closing in favor of #370

**Comment by npmccallum** [comment]
@connorkuehl Unless you have a specific objection, I'd like to keep this open and merge this onto master and rebase `demo2020` on top of this. I think we also probably have time now to address the other issues in the review. Thoughts?

**Comment by connorkuehl** [comment]
Discussed this briefly in gitter. I'm punting the refactor with a `Range` type until we have a tree with `vmsyscalls`, `enarx-keep-sev`, and `enarx-keep-sev-shim` all in one tree so that testing is easier and the replacement can be easily done all in one fell-swoop without affecting others' PRs.
